python/example_code/firehose/scenarios/firehose-put-actions/firehose.py

```diff
@@ -0,0 +1,253 @@
+# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+# SPDX-License-Identifier: Apache-2.0
+import json
+import logging
+import random
+from datetime import datetime, timedelta
+
+import backoff
+import boto3
+from config import get_config
+
+
+def load_sample_data(path: str) -> dict:
+    """
+    Load sample data from a JSON file.
+
+    Args:
+        path (str): The file path to the JSON file containing sample data.
+
+    Returns:
+        dict: The loaded sample data as a dictionary.
+    """
+    with open(path, "r") as f:
+        return json.load(f)
+
+
+# Configure logging
+logging.basicConfig(level=logging.INFO)
+logger = logging.getLogger(__name__)
+
+
+# snippet-start:[python.example_code.firehose.init]
+class FirehoseClient:
+    """
+    AWS Firehose client to send records and monitor metrics.
+
+    Attributes:
+        config (object): Configuration object with delivery stream name and region.
+        delivery_stream_name (str): Name of the Firehose delivery stream.
+        region (str): AWS region for Firehose and CloudWatch clients.
+        firehose (boto3.client): Boto3 Firehose client.
+        cloudwatch (boto3.client): Boto3 CloudWatch client.
+    """
+
+    def __init__(self, config):
+        """
+        Initialize the FirehoseClient.
+
+        Args:
+            config (object): Configuration object with delivery stream name and region.
+        """
+        self.config = config
+        self.delivery_stream_name = config.delivery_stream_name
+        self.region = config.region
+        self.firehose = boto3.client("firehose", region_name=self.region)
+        self.cloudwatch = boto3.client("cloudwatch", region_name=self.region)
+
+    # snippet-end:[python.example_code.firehose.init]
+
+    # snippet-start:[python.example_code.firehose.put_record]
+    @backoff.on_exception(
```

Comment:
The SDK has a backoff implementation built-in, does it not? I wondered why you use this third party backoff tool.

Reply:
It does not have backoff and jitter, just retry with some default backoff. It was recommended by the service-level team to add more configurable options for retry and backoff. The 'backoff` library is not a third-party tool but helps with easy implementation of backoff and jitter instead of writing it all yourself.